snuba/datasets/errors_replacer.py

```diff
@@ -667,14 +665,126 @@ def process_unmerge(
-        )
+        if convert_types:
+            return ", ".join(
+                "toUUID('%s')" % str(uuid.UUID(_hashify(h))) for h in hashes
+            )
+        else:
+            return ", ".join("'%s'" % str(uuid.UUID(_hashify(h))) for h in hashes)
     else:
-        query_args["hashes"] = ", ".join("'%s'" % _hashify(h) for h in hashes)
+        if convert_types:
+            return ", ".join("toFixedString('%s', 32)" % _hashify(h) for h in hashes)
+        else:
+            return ", ".join("'%s'" % _hashify(h) for h in hashes)
+
+
+def process_unmerge_hierarchical(
+    message: Mapping[str, Any],
+    all_columns: Sequence[FlattenedColumn],
+    state_name: ReplacerState,
+) -> Optional[Replacement]:
+    timestamp = datetime.strptime(message["datetime"], settings.PAYLOAD_DATETIME_FORMAT)
+    all_column_names = [c.escaped for c in all_columns]
+    select_columns = map(
+        lambda i: i if i != "group_id" else str(message["new_group_id"]),
+        all_column_names,
+    )
+
+    hierarchical_hashes = message.get("hierarchical_hashes") or {}
+    if not hierarchical_hashes:
+        return None
+
+    assert all(isinstance(h, str) for h in hierarchical_hashes.values())
+
+    prewhere = []
+    where = []
+    query_args = {}
+
+    # primary_hash is the root (most coarse) level of the grouping tree,
+    # meaning:
+    #
+    # 1. primary_hash == hierarchical_hashes[0]
+    # 2. <any entry in hierarchical_hashes>:primary_hash is a n:1 relation
+    #
+    # Sentry wants us to filter events by hierarchical_hashes entry using
+    # hasAny, but also tells us the corresponding primary_hash such that we
+    # can filter more efficiently.
+    #
+    # For now this is just converted into
+    #
+    #     PREWHERE primary_hash IN ... WHERE hasAny(hierarchical_hashes, ...)
+    #
+    # ..but perhaps one day we may decide to instead emit multiple replacements of the form
+    #
+    #     PREWHERE primary_hash = ... WHERE has(hierarchical_hashes, ...)
+    #
+    # ..if we think that it would be faster.
+    prewhere.append("primary_hash IN (%(primary_hashes_for_hierarchical_hashes)s)")
+    query_args["primary_hashes_for_hierarchical_hashes"] = _convert_hash_list(
+        list(set(hierarchical_hashes.values())), state_name
+    )
+
+    where.append("hasAny(hierarchical_hashes, [%(hierarchical_hashes)s])")
+    query_args["hierarchical_hashes"] = _convert_hash_list(
+        list(hierarchical_hashes), state_name, convert_types=True
+    )
+
+    # group_id cannot be part of PREWHERE because FINAL is applied before
+    # PREWHERE, meaning if group_id was part of prewhere we would see events
+    # that are no longer part of this group.
+    #
+    # As long as group_id is not in PREWHERE, the query result is guaranteed to
+    # be constrained to one group. primary_hash can now be in PREWHERE, if that
+    # were not OK it would mean we have a replacement row in storage that did
```

Comment:
why would it be not ok ?

Reply:
I'm trying proof-by-contradiction here. "assuming this were not OK, it would mean X which is obviously incorrect"